cmd/kubeadm/app/util/marshal_test.go

```diff
@@ -111,50 +107,48 @@ func TestMarshalUnmarshalYaml(t *testing.T) {
 	}
 }
 
-func TestMarshalUnmarshalToYamlForCodecs(t *testing.T) {
-	cfg := &kubeadmapiv1.InitConfiguration{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       constants.InitConfigurationKind,
-			APIVersion: kubeadmapiv1.SchemeGroupVersion.String(),
-		},
-		NodeRegistration: kubeadmapiv1.NodeRegistrationOptions{
-			Name:      "testNode",
-			CRISocket: "unix:///var/run/cri.sock",
-		},
-		BootstrapTokens: []bootstraptokenv1.BootstrapToken{
-			{
-				Token: &bootstraptokenv1.BootstrapTokenString{ID: "abcdef", Secret: "abcdef0123456789"},
-			},
-		},
-		// NOTE: Using MarshalToYamlForCodecs and UnmarshalFromYamlForCodecs for ClusterConfiguration fields here won't work
-		// by design. This is because we have a `json:"-"` annotation in order to avoid struct duplication. See the comment
-		// at the kubeadmapiv1.InitConfiguration definition.
+func TestUnmarshalJson(t *testing.T) {
+	bytes := []byte(string(`{
+  "apiVersion": "v1",
+  "kind": "Pod",
+  "metadata": {
+    "name": "someName",
+    "namespace": "testNamespace",
+	"labels": {
+		"test": "yes"
 	}
+  },
+  "spec": {
+	"restartPolicy": "Always"
+  }
+}`))
 
-	kubeadmapiv1.SetDefaults_InitConfiguration(cfg)
-	scheme := runtime.NewScheme()
-	if err := kubeadmapiv1.AddToScheme(scheme); err != nil {
-		t.Fatal(err)
-	}
-	codecs := serializer.NewCodecFactory(scheme)
+	t.Logf("\n%s", bytes)
 
-	bytes, err := MarshalToYamlForCodecs(cfg, kubeadmapiv1.SchemeGroupVersion, codecs)
+	obj2, err := UniversalUnmarshal(bytes)
 	if err != nil {
-		t.Fatalf("unexpected error marshalling InitConfiguration: %v", err)
+		t.Fatalf("unexpected error marshalling: %v", err)
 	}
-	t.Logf("\n%s", bytes)
 
-	obj, err := UnmarshalFromYamlForCodecs(bytes, codecs)
-	if err != nil {
-		t.Fatalf("unexpected error unmarshalling InitConfiguration: %v", err)
+	pod2, ok := obj2.(*corev1.Pod)
+	if !ok {
+		t.Fatal("did not get a Pod")
+	}
+
+	if pod2.Name != "someName" {
+		t.Errorf("expected someName, got %q", pod2.Name)
+	}
+
+	if pod2.Namespace != "testNamespace" {
+		t.Errorf("expected testNamespace, got %q", pod2.Namespace)
 	}
 
-	cfg2, ok := obj.(*kubeadmapiv1.InitConfiguration)
-	if !ok || cfg2 == nil {
-		t.Fatal("did not get InitConfiguration back")
+	if !reflect.DeepEqual(pod2.Labels, map[string]string{"test": "yes"}) {
+		t.Errorf("expected [test:yes], got %v", pod2.Labels)
 	}
-	if !reflect.DeepEqual(*cfg, *cfg2) {
-		t.Errorf("expected %v, got %v", *cfg, *cfg2)
+
+	if pod2.Spec.RestartPolicy != "Always" {
+		t.Errorf("expected Always, got %q", pod2.Spec.RestartPolicy)
```

Comment:
Add a false check here, e.g. if there is something default from the pod found, might be `pod.Spec.DNSPolicy` or `.Spec.Containers[0].TerminationMessagePath`, then error it!

Reply:
I thought this is covered in another UT of `TestCreateLocalEtcdStaticPodManifestFileWithPatches`. So I did not add the check logic here. Adding the check makes sense as well.